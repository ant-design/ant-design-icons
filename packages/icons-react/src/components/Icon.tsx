import * as React from 'react';
import classNames from 'classnames';

import { svgBaseProps, warning, useInsertStyles } from '../utils';

export interface IconBaseProps extends React.HTMLProps<HTMLSpanElement> {
  spin?: boolean;
  rotate?: number;
}

export interface CustomIconComponentProps {
  width: string | number;
  height: string | number;
  fill: string;
  viewBox?: string;
  className?: string;
  style?: React.CSSProperties;
}

export interface IconComponentProps extends IconBaseProps {
  viewBox?: string;
  component?: React.ComponentType<CustomIconComponentProps | React.SVGProps<SVGSVGElement>>;
  ariaLabel?: React.AriaAttributes['aria-label'];
}

const Icon = React.forwardRef<HTMLSpanElement, IconComponentProps>((props, ref) => {
  const {
    // affect outter <i>...</i>
    className,

    // affect inner <svg>...</svg>
    component: Component,
    viewBox,
    spin,
    rotate,

    tabIndex,
    onClick,

    // children
    children,
    ...restProps
  } = props;

  warning(
    Boolean(Component || children),
    'Should have `component` prop or `children`.',
  );

  useInsertStyles();

  const classString = classNames(
    'anticon',
    className,
  );

  const svgClassString = classNames({
    'anticon-spin': !!spin,
  });

  const svgStyle = rotate
    ? {
        msTransform: `rotate(${rotate}deg)`,
        transform: `rotate(${rotate}deg)`,
      }
    : undefined;

  const innerSvgProps: CustomIconComponentProps = {
    ...svgBaseProps,
    className: svgClassString,
    style: svgStyle,
    viewBox,
  };

  if (!viewBox) {
    delete innerSvgProps.viewBox;
  }

  // component > children
  const renderInnerNode = () => {
    if (Component) {
      return <Component {...innerSvgProps}>{children}</Component>;
    }

    if (children) {
      warning(
        Boolean(viewBox) ||
          (React.Children.count(children) === 1 &&
            React.isValidElement(children) &&
            React.Children.only(children).type === 'use'),
        'Make sure that you provide correct `viewBox`' +
        ' prop (default `0 0 1024 1024`) to the icon.',
      );

      return (
        <svg {...innerSvgProps} viewBox={viewBox}>
          {children}
        </svg>
      );
    }

    return null;
  }

  let iconTabIndex = tabIndex;
  if (iconTabIndex === undefined && onClick) {
    iconTabIndex = -1;
  }

  return (
    <span
      role="img"
      {...restProps}
      ref={ref}
      tabIndex={iconTabIndex}
      onClick={onClick}
      className={classString}
    >
      {renderInnerNode()}
    </span>
  );
});

Icon.displayName = 'AntdIcon';

export default Icon;
