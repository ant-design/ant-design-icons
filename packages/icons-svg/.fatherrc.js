module.exports = {
  esm: {
    type: 'babel'
  },
  cjs: {
    type: 'babel'
  }
  // disableTypeCheck: true
};
