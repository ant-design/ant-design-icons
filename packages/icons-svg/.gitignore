.DS_Store
node_modules/
**/*/node_modules/
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# Editor directories and files
.idea
.vscode
*.suo
*.ntvs*
*.njsproj
*.sln

# lock
package-lock.json
yarn.lock

# rollup cache
.rpt2_cache

# auto generated
_site/
.cache
public
src
inline-svg
!docs/inline-svg
es
lib

# docz
.docz
docz-dist

# publish
*.tgz
