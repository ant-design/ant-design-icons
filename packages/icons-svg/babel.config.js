// for test env only

module.exports = {
  presets: [
    [
      '@babel/preset-env',
      {
        targets: {
          node: 'current'
        }
      }
    ],
    '@babel/preset-typescript'
  ]
};
