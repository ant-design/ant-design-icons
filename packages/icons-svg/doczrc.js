export default {
  dest: '/docz-dist',
  typescript: true
};
