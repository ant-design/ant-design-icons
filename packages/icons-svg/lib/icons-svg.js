'use strict';

module.exports = iconsSvg;

function iconsSvg() {
    // TODO
}
