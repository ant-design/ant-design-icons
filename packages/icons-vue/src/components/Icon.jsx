import { svgBaseProps, warning, useInsertStyles } from '../utils';

const Icon = {
  name: 'AntdIcon',
  functional: true,
  props: ['component', 'spin', 'rotate', 'tabIndex'],
  render(h, ctx) {
    const { data: { attrs, ...restData } = {}, props = {}, listeners, children } = ctx;
    const {
      // affect inner <svg>...</svg>
      component: Component,
      viewBox,
      spin,
      rotate,
      tabIndex,
      ...restProps
    } = { ...attrs, ...props };
    const { click: onClick } = listeners;
    const hasChildren = children && children.length;
    warning(Boolean(Component || hasChildren), 'Should have `component` prop or `children`.');

    useInsertStyles();

    const classString = 'anticon';

    const svgClassString = {
      'anticon-spin': !!spin,
    };
    const svgStyle = rotate
      ? {
          msTransform: `rotate(${rotate}deg)`,
          transform: `rotate(${rotate}deg)`,
        }
      : undefined;

    const innerSvgProps = {
      attrs: { ...svgBaseProps, viewBox },
      class: svgClassString,
      style: svgStyle,
    };

    if (!viewBox) {
      delete innerSvgProps.attrs.viewBox;
    }
    const renderInnerNode = () => {
      // component > children
      if (Component) {
        return typeof Component === 'function' ? (
          Component(h, { ...innerSvgProps, children })
        ) : (
          <Component {...innerSvgProps}>{children}</Component>
        );
      }
      if (hasChildren) {
        warning(
          Boolean(viewBox) || (children.length === 1 && children[0] && children[0].tag === 'use'),
          'Make sure that you provide correct `viewBox`' +
            ' prop (default `0 0 1024 1024`) to the icon.',
        );

        return (
          <svg {...innerSvgProps} viewBox={viewBox}>
            {children}
          </svg>
        );
      }
      return null;
    };

    let iconTabIndex = tabIndex;
    if (iconTabIndex === undefined && onClick) {
      iconTabIndex = -1;
    }

    return (
      <span
        role="img"
        tabIndex={iconTabIndex}
        class={classString}
        {...{ ...restData, attrs: restProps, on: listeners }}
      >
        {renderInnerNode()}
      </span>
    );
  },
};

export default Icon;
