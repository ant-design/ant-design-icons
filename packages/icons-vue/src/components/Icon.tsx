import { svgBaseProps, warning } from '../utils';
import { Component, FunctionalComponent, HTMLAttributes, PropType } from 'vue';
import { useInjectIconContext } from './Context';
import { InsertStyles } from './InsertStyle';

export interface IconBaseProps extends HTMLAttributes {
  spin?: boolean;
  rotate?: number;
}
export interface CustomIconComponentProps {
  width: string | number;
  height: string | number;
  fill: string;
  viewBox?: string;
}
export interface IconComponentProps extends IconBaseProps {
  viewBox?: string;
  component?: Component;
  ariaLabel?: string;
}
export interface IconType extends FunctionalComponent<IconComponentProps> {
  displayName: string;
}
const Icon: IconType = (props, context) => {
  const { attrs, slots } = context;
  const {
    class: cls,
    // affect inner <svg>...</svg>
    component: Component,
    viewBox,
    spin,
    rotate,
    tabindex,
    onClick,
    ...restProps
  } = { ...props, ...attrs } as any;
  const { prefixCls, rootClassName } = useInjectIconContext();
  const children = slots.default && slots.default();
  const hasChildren = children && children.length;
  const slotsComponent = slots.component;
  warning(
    Boolean(Component || hasChildren || slotsComponent),
    'Should have `component` prop/slot or `children`.',
  );

  const classString = {
    [rootClassName.value]: !!rootClassName.value,
    [prefixCls.value]: true,
  };

  const svgClassString = {
    [`${prefixCls.value}-spin`]: spin === '' || !!spin,
  };
  const svgStyle = rotate
    ? {
        msTransform: `rotate(${rotate}deg)`,
        transform: `rotate(${rotate}deg)`,
      }
    : undefined;

  const innerSvgProps = {
    ...svgBaseProps,
    viewBox,
    class: svgClassString,
    style: svgStyle,
  };

  if (!viewBox) {
    delete innerSvgProps.viewBox;
  }
  const renderInnerNode = () => {
    if (Component) {
      return <Component {...innerSvgProps}>{children}</Component>;
    }
    if (slotsComponent) {
      return slotsComponent(innerSvgProps);
    }
    if (hasChildren) {
      warning(
        Boolean(viewBox) || (children.length === 1 && children[0] && children[0].type === 'use'),
        'Make sure that you provide correct `viewBox`' +
          ' prop (default `0 0 1024 1024`) to the icon.',
      );

      return (
        <svg {...innerSvgProps} viewBox={viewBox}>
          {children}
        </svg>
      );
    }
    return null;
  };

  let iconTabIndex = tabindex;
  if (iconTabIndex === undefined && onClick) {
    iconTabIndex = -1;
    restProps.tabindex = iconTabIndex;
  }

  return (
    <span role="img" {...restProps} onClick={onClick} class={[classString, cls]}>
      {renderInnerNode()}
      <InsertStyles></InsertStyles>
    </span>
  );
};
Icon.props = {
  spin: Boolean as PropType<boolean>,
  rotate: Number as PropType<number>,
  viewBox: String as PropType<string>,
  ariaLabel: String as PropType<string>,
};
Icon.inheritAttrs = false;
Icon.displayName = 'Icon';

export default Icon;
