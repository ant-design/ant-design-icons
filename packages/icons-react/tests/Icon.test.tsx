import { TwitterOutlined } from '@ant-design/icons-svg';
import * as React from 'react';
import { mount } from 'enzyme';
import Icon from '../src/components/IconBase';

describe('React AntdIcon Component', () => {
  it('should create SVG element.', () => {
    const props = {
      style: {
        fontSize: '3rem',
      },
      className: 'my-icon',
      extraProps: { hello: 'world' },
    } as any;
    const icon = mount(
      <Icon {...props} />,
    );

    expect(icon).toMatchSnapshot();
  });

  it('should allow explicit import.', () => {
    const icon = mount(<Icon icon={TwitterOutlined} />);
    expect(typeof icon.find('path').first().prop('d')).toBe('string');
  });

  it('should render null, when the type is invalid.', () => {
    const iconWithObjectTypeProp = mount(<Icon icon={{ invalid: true } as any} />);
    expect(iconWithObjectTypeProp.isEmptyRender()).toBeTruthy();

    const iconWithStringTypeProp = mount(<Icon icon={'Later is better than never.' as any} />);
    expect(iconWithStringTypeProp.isEmptyRender()).toBeTruthy();
  });
});
