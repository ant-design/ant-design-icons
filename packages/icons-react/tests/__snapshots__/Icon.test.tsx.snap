// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React AntdIcon Component should create SVG element. 1`] = `
<IconReact
  className="my-icon"
  extraProps={
    Object {
      "hello": "world",
    }
  }
  style={
    Object {
      "fontSize": "3rem",
    }
  }
/>
`;
